### PR TITLE
Json Marshaller Migration To Poet

### PR DESCRIPTION
**Pull request by zoewangg**
## Description
Related to #30 
To favor our new practices of making PR smaller and getting reviewed faster. This task is broken into the the following parts:

1. Marshaller Refactor 
   a. Json, rest-json marshaller
   b. query xml, rest-xml, ec2 marshaller
   c. api gateway marshaller

1. Unmarshaller Refacctor
   a. Json, rest-json marshaller
   b. query xml, rest-xml, ec2 marshaller
   c. api gateway marshaller 

This PR is for `1.a Json, rest-json marshaller`
## Motivation and Context
- Replacing freemarker with poet
- Removing unused files.

## Testing
`mvn clean install` and integration test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [] A short description of the change has been added to the **CHANGELOG**

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/transform/MarshallerSpec.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import software.amazon.awssdk.codegen.poet.PoetExtensions;
+import software.amazon.awssdk.codegen.poet.PoetUtils;
+import software.amazon.awssdk.codegen.poet.transform.protocols.JsonMarshallerSpec;
+import software.amazon.awssdk.codegen.poet.transform.protocols.MarshallerProtocolSpec;
+import software.amazon.awssdk.core.Request;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.runtime.transform.Marshaller;
+import software.amazon.awssdk.utils.Validate;
+
+public class MarshallerSpec implements ClassSpec {
```
Comment (shorea) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/transform/MarshallerSpec.java]: Not exactly sure what this is for.

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/operationwithnoinputoroutputrequestmodelmarshaller.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    /**
+     * Marshall the given parameter object
+     */
+    public void marshall(OperationWithNoInputOrOutputRequest operationWithNoInputOrOutputRequest,
+                         ProtocolMarshaller protocolMarshaller) {
+        Validate.paramNotNull(operationWithNoInputOrOutputRequest, "operationWithNoInputOrOutputRequest");
+        Validate.paramNotNull(protocolMarshaller, "protocolMarshaller");
+        try {
```
Comment (shorea) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/operationwithnoinputoroutputrequestmodelmarshaller.java]: Be nice to get rid of this if it's easy to do.

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/streaminginputoperationrequestmodelmarshaller.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        return INSTANCE;
+    }
+
+    /**
+     * Marshall the given parameter object
+     */
+    public void marshall(StreamingInputOperationRequest streamingInputOperationRequest, ProtocolMarshaller protocolMarshaller) {
+        Validate.paramNotNull(streamingInputOperationRequest, "streamingInputOperationRequest");
+        Validate.paramNotNull(protocolMarshaller, "protocolMarshaller");
+        try {
```
Comment (shorea) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/streaminginputoperationrequestmodelmarshaller.java]: This doesn't seem right, shouldn't it be marshalling the streaming member?

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/transform/JsonModelMarshallerSpec.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+    @Override
+    public TypeSpec poetSpec() {
+
+        TypeSpec.Builder specBuilder = TypeSpec.classBuilder(className)
+                                               .addModifiers(Modifier.PUBLIC)
+                                               .addAnnotation(PoetUtils.GENERATED)
+                                               .addAnnotation(SdkInternalApi.class)
+                                               .addMethods(marshalMethods())
+                                               .addFields(memberVariables())
```
[the remark this answers is not in the document]
Comment (kiiadi) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/transform/JsonModelMarshallerSpec.java]: I'm not sure that poet does reorder annotations - I'm fairly sure they'll go on in the order that they're called on the builder....

[the remark this answers is not in the document]
Comment (kiiadi) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/transform/JsonModelMarshallerSpec.java]: Or did you mean fields/methods (that it does do correctly)

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/transform/JsonModelMarshallerSpec.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    public ClassName className() {
+        return className;
+    }
+
+    private List<MethodSpec> marshalMethods() {
+        List<MethodSpec> methodSpecs = new ArrayList<>();
+
+        methodSpecs.add(constructor());
+        methodSpecs.add(getInstanceMethod());
+        methodSpecs.add(marshalMethod());
```
[the remark this answers is not in the document]
Comment (kiiadi) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/transform/JsonModelMarshallerSpec.java]: We should definitely choose one and be consistent!

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/transform/JsonModelMarshallerSpec.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (shorea) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/transform/JsonModelMarshallerSpec.java]: yeah fields and methods, here the member variables are added after the method but poet rearranges it. just looked weird when reading it

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/transform/MarshallerSpec.java (hunk 1 above)
Comment (zoewangg) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/transform/MarshallerSpec.java]: This is the high level class used in`MarshallerGeneratorTasks` to create the marshaller class and it uses different implementations of `MarshallerProtocolSpec` to add the marshaller methods based on the protocol. This PR only contains `JsonMarshallerSpec`. This class is  similar to [AsyncClientClass](https://github.com/aws/aws-sdk-java-v2/blob/master/codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/AsyncClientClass.java)

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/operationwithnoinputoroutputrequestmodelmarshaller.java (hunk 2 above)
Comment (zoewangg) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/operationwithnoinputoroutputrequestmodelmarshaller.java]: good point

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/streaminginputoperationrequestmodelmarshaller.java (hunk 3 above)
Comment (zoewangg) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/transform/streaminginputoperationrequestmodelmarshaller.java]: don't we use [StreamingRequestMarshaller](https://github.com/aws/aws-sdk-java-v2/blob/master/core/src/main/java/software/amazon/awssdk/core/runtime/transform/StreamingRequestMarshaller.java) to marshall all streaming members in v2?

**Pull request merged**